File: 2021/csharp/Aoc2021.Tests/Day02Tests.cs
namespace Aoc2021.Tests;

public class Day02Tests
{
    private static readonly List<Day02.Command> commands = Day02.ParseCommands(new[]
    {
        "forward 5",
        "down 5",
        "forward 8",
        "up 3",
        "down 8",
        "forward 2",
    });

    [Fact]
    public void MovingPart1Works()
    {
        Assert.Equal(new(15, 10, 0), Day02.MoveFirstTry(new(0, 0, 0), commands));
    }

    [Fact]
    public void MovingPart2Works()
    {
        Assert.Equal(new(15, 60, 0), Day02.MoveSecondTry(new(0, 0, 0), commands) with { Aim = 0 });
    }

    [Fact]
    public void Stars()
    {
        Day02 run = new();
        Assert.Equal("1507611", run.Result1());
        Assert.Equal("1880593125", run.Result2());
    }
}

File: 2021/fsharp/Aoc2021.Tests/Day02.Tests.fs
module Day02.Tests

open Xunit
open Day02

let commands =
    [ "forward 5"; "down 5"; "forward 8"; "up 3"; "down 8"; "forward 2" ]
    |> parseCommands

let submarine =
    { HorizontalPosition = 0
      Depth = 0
      Aim = 0 }

[<Fact>]
let ``Day02 move part 1 works`` () =
    Assert.Equal(
        { HorizontalPosition = 15
          Depth = 10
          Aim = 0 },
        commands |> moveFirstTry submarine
    )

[<Fact>]
let ``Day02 move part 2 works`` () =
    Assert.Equal(
        { HorizontalPosition = 15
          Depth = 60
          Aim = 0 },
        { (commands |> moveSecondTry submarine) with
            Aim = 0 }
    )

[<Fact>]
let ``Day02 Stars`` () =
    Assert.Equal("1507611", job1 ())
    Assert.Equal("1880593125", job2 ())
